doc/contrib/new_api.md

```diff
@@ -0,0 +1,510 @@
+# PREVIEW: New `libelektra-core` API 2023
+
+## Core API
+
+### Note on names
+
+From the [use cases](/doc/usecases/core/README.md) we know that there is a collection structure which contains elements with a name and value (and metadata).
+
+The current API uses these names for those concepts:
+
+- Collection: `KeySet`
+- Element: `Key`
+- Name of element: —
+- Value of element: —
+
+The API proposed below uses (to keep as much as possible from above):
+
+- Collection: `ElektraKeyset`
+- Element: `ElektraKey`
+- Name of element: `ElektraKeyname`
+- Value of element: `ElektraKeyvalue`
+
+Another (maybe less confusing) option would be:
+
+- Collection: `ElektraSet` or `ElektraMap`
```

Comment:
:+1: ElektraSet is nearer to current name, I don't see a reason to change to Map.

Reply:
Well "Map, Entry, Key, Value" would be a pretty standard combination of terms.

Although changing the meaning of "Key" like that will be confusing for a while. It also means we actually need to update **all the docs** in more or less one go, instead of adding a notice "we changed terms from X to Y" somewhere and updating over a longer period of time.

If we want to use "Set", I actually prefer the combination: "Set, Entry, Name, Value".

Reply:
> instead of adding a notice "we changed terms from X to Y" 

This doesn't work anyway (even with FLOSS it would never be finished).

> If we want to use "Set", I actually prefer the combination: "Set, Entry, Name, Value".

Probably better to avoid `Key` anyway, see https://github.com/ElektraInitiative/libelektra/pull/4805#discussion_r1066061990

Reply:
> This doesn't work anyway (even with FLOSS it would never be finished).

It is not intended as a permanent solution. But exactly because of FLOSS/CM having such a notice might be useful. Otherwise people could get confused, while the change happens, or we are limited to making the change when there is no FLOSS/CM happening.

Also we could always accidentally miss an occurrence in the update. The notice could also be put into the release notes. If we reuse "Key" we end up with something like:

```
Key was renamed to Entry and Keyname was renamed to Key.
```

Which is ambiguous. You can't tell whether "Key" is the old meaning and we forgot to update, or it is the new meaning.

Reply:
Back to topic: can you maybe create a second PR with this new naming schema? Maybe the look&feel gives a clearer picture.

I think "Key" -> "Entry" is a bit like our ref counting: a historical "problem", where the "problem" is that we do it differently then the rest of the world.

So these are border-line decisions (not very convincing problem but also not totally unconvincing): we should only tackle them if you are sure you can actually fully migrate&fix everything.

Reply:
> we should only tackle them if you are sure you can actually fully migrate&fix everything

Renaming in C code is not an issue. Renaming in most other languages should be doable too. The issue is with documentation. I fear that has to be done manually. 

> I think "Key" -> "Entry" is a bit like our ref counting:

The difference is
1. ref counting is a technical issue with pros and cons on both sides
2. ref counting is a somewhat advanced topic that most people shouldn't need to worry about.

Using "Key" for something that is clearly not just a key, meanwhile, is just plain confusing to people who don't already know.

> Can you maybe create a second PR with this new naming schema?

I'll create a self-contained commit in this PR. If we like it we keep, if not I'll just revert the commit. Creating a new PR, unnecessarily splits the conversation. (It's unlikely the discussion in the new PR would really just focus on the names and nothing else).